### PR TITLE
add `OpInfo` for `torch.nn.functional.kl_div`

### PR DESCRIPTION
**Pull request by pmeier**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #65681
* #68419
* #69212
* #67079
* #67023
* #67022
* #69211
* #65680
* __->__ #65469

Differential Revision: [D31111698](https://our.internmc.facebook.com/intern/diff/D31111698)

cc @albanD @mruberry @jbschlosser @walterddr @ezyang @gchanan

### CODE REVIEW

**Comment by pmeier** [comment]
Failures are expected. PR is blocked until #65466 is resolved.

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31111698).

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31111698).

**Comment by pmeier** [comment]
The latest commit cleans up the implementation of `kl_div` by quite a bit, which ultimately also resolved #65466:

- remove the split between `kl_div_backward_cpu` and `kl_div_backward_cuda`
- remove the explicit double backward

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (pmeier) [inline torch/csrc/autograd/FunctionsManual.cpp:1180]: The `masked_fill` is actually wrong here. While `x * log(x)` for `x == 0` is commonly defined as `0`, its gradient `1 + log(x)` is `-inf` there.

Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:1180]: Nit. It's not "commonly defined as 0, but it so happens that lim x -> 0 xlog(x) = 0.

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (Lezcano) [inline torch/csrc/autograd/FunctionsManual.cpp:1167]: You can probably delete this line.

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (Lezcano) [inline torch/csrc/autograd/FunctionsManual.cpp:1181]: Same as above, you can probably delete this line.

File: torch/testing/_internal/common_methods_invocations.py
Comment (Lezcano) [inline torch/testing/_internal/common_methods_invocations.py:5685]: Given that the `log_target` takes different routes in the backward, might be good to test the 8 combination log_target x reduction?

File: aten/src/ATen/core/Reduction.h
Comment (Lezcano) [inline aten/src/ATen/core/Reduction.h:13]: Is it fine to add this here without it being supported by any of the other losses that accept a reduction? What happens if we pass `"batchmean"` as the reduction to any of the other losses?

**Review by lezcano — COMMENTED** [review]
This looks good, I just left a few comments.

File: aten/src/ATen/core/Reduction.h
Comment (pmeier) [reply aten/src/ATen/core/Reduction.h:13]: You are right, that shouldn't be there. It is handled completely on the Python side:

https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L2801-L2803

I wonder where the name `"batchmean"` came from if its apparently the same as `"sum"`.

File: torch/testing/_internal/common_nn.py
Comment (pmeier) [inline torch/testing/_internal/common_nn.py:516]: The target of `KLDivLoss` can either be a probability `[0, 1]` or a log-probability `(-inf, 0]`. I've fixed the sample inputs here, but there are some more failures due to 

https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/testing/_internal/common_nn.py#L5709

in tests that are automatically generated. Not sure what to do about them.

**Comment by ezyang** [comment]
Backwards removal is technically BC breaking, marking as such. (I'm not too sure if it should be removed here or not, hopefully someone who is more familiar here can make the call.)

File: torch/testing/_internal/common_nn.py
Comment (pmeier) [reply torch/testing/_internal/common_nn.py:516]: The failures are semi-legit:

- they actually show a divergence between the forward and backward pass:
- the divergence is caused by a really odd choice in the forward pass https://github.com/pytorch/pytorch/blob/6a990535154116960767ad422e724250ecfb8325/aten/src/ATen/native/Loss.cpp#L114-L117
  This silently ignores negative inputs in case of a non-log target. Not only is this not documented, but also indicates a bug in the user pipeline. 

Although this would be BC breaking, IMO we should remove this special case for negative values. Of course we still need to keep the continuous extension for `0 * log(0) := 0` in the forward pass.

Comment (Lezcano) [reply torch/testing/_internal/common_nn.py:516]: So, the KL divergence is just defined for for probabilities in [0,1]. Even more, there's nothing stopping you from defining the KL-divergence for general measures, so you could think that the inputs are in [0, infty]. Now, enforcing this would mean checking every input which requires some extra computations.

In my opinion, I think it is good to not check that the inputs are positive, and if they are not, you'll end up returning a `-inf`. If this is well-documented and tested, I think it's better than an arbitrary choice of `0` that's not documented.

Now, this is BC-breaking, but it's BC-breaking with respect to an undocumented "feature", so I believe we should be free to fix these things. cc @jbschlosser

File: torch/testing/_internal/common_nn.py
Comment (pmeier) [inline torch/testing/_internal/common_nn.py:4677]: This is an ugly hack, but I feel like this whole file is. Basically, this is a one-liner that is able to create `nn.KLDivLoss` with `log_target=True` and also have `.__name__ == "KLDivLoss"`. cc @thomasjpfan @jbschlosser

File: torch/testing/_internal/common_nn.py
Comment (pmeier) [inline torch/testing/_internal/common_nn.py:5476]: We need to exclude `KLDivLoss` here, because just setting the input and target sizes (L5483-L5484) will create the inputs with `randn` (see https://github.com/pytorch/pytorch/pull/65469/files#r716708268). This would result in NaN's which would fail the gradcheck.

**Review by lezcano — APPROVED** [review]
Just left a couple small comments. Everything else looks good modulo making the CI go green.

cc @JackCaoG regarding the XLA failure due to moving `kl_backward` to `FunctionsManual.cpp`

**Comment by pmeier** [comment]
@JackCaoG What is the status on the XLA side?

**Comment by JackCaoG** [comment]
@pmeier working on https://github.com/pytorch/xla/pull/3185, fixing one last test. Should be good to go pretty soon.

**Comment by JackCaoG** [comment]
@pmeier xla pr is ready

**Comment by pmeier** [comment]
Ping @mruberry @jbschlosser for review.

File: aten/src/ATen/native/Loss.cpp
Comment (mruberry) [inline aten/src/ATen/native/Loss.cpp:111]: This is technically a different function than what was previously implemented, right? Would you elaborate on the change here?

File: aten/src/ATen/native/native_functions.yaml
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:2386]: I don't think anyone is using this but this is, of course, a BC-breaking change.

File: test/onnx/test_pytorch_onnx_onnxruntime.py
Comment (mruberry) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:7047]: Nice ONNX fixes

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (mruberry) [inline torch/csrc/autograd/FunctionsManual.cpp:1280]: Would you elaborate on the changes here, too?

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (mruberry) [inline torch/csrc/autograd/FunctionsManual.cpp:1269]: Isn't this going to be a performance regression over the previous implementation because previously there were handwritten fused kernels for these operations?

**Review by mruberry — COMMENTED** [review]
Wow there is a lot going on here!

Let's separate the OpInfo, which looks great, from the other changes. For the other changes I have some concerns that I added inline comments for. Looking forward to discussing them!

File: aten/src/ATen/native/Loss.cpp
Comment (Lezcano) [reply aten/src/ATen/native/Loss.cpp:111]: It is indeed different in that the KL of a negative number this implementation returns `NaN`, while before it returned zero. This is consistent with the formula. Even more, it is consistent with the idea that the function is not defined for negative numbers.
On a side note, it's also faster to do this than to return zeros.

Comment (Lezcano) [reply aten/src/ATen/native/Loss.cpp:111]: This is also consistent with the current state of the docs added here: https://github.com/pytorch/pytorch/pull/67443

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:1280]: Same thing as in the other case, it concerns negative numbers, for which this function is not defined. It is also faster to do it this way, and it's also consistent with the docs.

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:1269]: Potentially, although just for CUDA and for `log_target=False` as the TI was just on that branch. It should be faster on the CPU side though.

I would not worry too much about this, as, in that case, just 3 kernels are being called. I reckon that this is not such a big deal (?)

**Comment by pmeier** [comment]
After some offline discussion with @mruberry and @Lezcano, we opted to separate the BC breaking part into a another PR (#68419), but keep everything else as is.

File: aten/src/ATen/native/cuda/Loss.cu
Comment (Lezcano) [inline aten/src/ATen/native/cuda/Loss.cu:49]: These function and the `_cpu` counterpart you should keep and remove in the other PR.

File: aten/src/ATen/native/native_functions.yaml
Comment (Lezcano) [inline aten/src/ATen/native/native_functions.yaml:2390]: Should leave as it was and just remove the whole entry in the next PR.

File: tools/autograd/derivatives.yaml
Comment (Lezcano) [inline tools/autograd/derivatives.yaml:1978]: In `derivatives.yaml` you'll need to change the derivative of `kl_div` to be `kl_div_backward_aux`, and then rename it to `kl_div_backward` in the next PR.

**Review by lezcano — APPROVED** [review]
LGTM if the CI agrees!

**Comment by mruberry** [comment]
Unfortunately @jbschlosser is under the weather at the moment but he should be better soon and we'll take a look

**Review by mruberry — COMMENTED** [review]
Hey @pmeier, @Lezcano! Thanks for your patience.

@mruberry and @jbschlosser here (@jbschlosser is feeling much better!). Here's our take on this PR:

- it's cool, but...
- we don't want to change the semantics of kl_div in this PR
- would you open an issue so we can discuss the NaN semantics more? 
- the test changes to ensure the inputs to kl_div are valid are awesome
- the OpInfo is cool, of course

So basically we'd like to shrink the PR's semantic changes and open an issue to discuss them separately. What are your thoughts?

**Comment by pmeier** [comment]
> So basically we'd like to shrink the PR's semantic changes and open an issue to discuss them separately. What are your thoughts?

This PR adds two changes to `kl_div`:

1. Entangles the logic, which should result in a speed-up since we are calling fewer kernels.
2. Removes undocumented behavior, which silently accept wrong inputs. Without `log_target=True`, `target` should be a probability in `[0.0, 1.0]`. The current implementation simply substitutes `0` for the loss in case it encounters a negative `target`:

    https://github.com/pytorch/pytorch/blob/fbaa19a6fa6c7d07ecd74e5ebf3c5d0fc77ecc51/aten/src/ATen/native/Loss.cpp#L117

    For example:
    
    ```python
    import torch
    from torch.nn.functional import kl_div
    
    input = torch.rand(()).log()
    target = torch.tensor(-1, dtype=input.dtype)
    
    print(kl_div(input, target))
    ```
    ```
    tensor(0.)
    ```

    This is problematic for two reasons:

    1. It is mathematically wrong, since the equation (in particular `log(target)`) is undefined for negative targets.
    2. It most likely hides a bug in the user code.

    Running the snippet above with the fixes in this PR gives `tensor(nan)` which resolves both issues.

Do you really want me to revert 2. to the old behavior?

**Comment by Lezcano** [comment]
I think what @pmeier mentioned is on point. The new implementation is faster and correct (follows the new docs we wrote for this op).
This looks like a similar case to that of `nll_loss`, only that in this case it is clear what the value should be. If the user passes a negative value to this function with `log_target=True` then that should return `NaN` by definition. 
cc @ngimel to chip in as well.

**Comment by pmeier** [comment]
After some more offline discussion, we decided to split all changes to the op into a different PR #69212. Thus, this PR no longer fixes #65466.

File: torch/testing/_internal/common_methods_invocations.py
Comment (saketh-are) [inline torch/testing/_internal/common_methods_invocations.py:7343]: Should we consider adding samples which pass `size_average` or `reduce` kwargs for completeness?

Comment (pmeier) [reply torch/testing/_internal/common_methods_invocations.py:7343]: I'd rather not. They are deprecated since `torch==0.4.1`, which was released over 2 years ago:

https://github.com/pytorch/pytorch/blob/a24163a95edb193ff7b06e98cd69bf7cfd4c0d2f/torch/nn/functional.py#L38

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:7519]: Let's add a comment explaining log target vs not and why we want the range we do for each case

**Review by mruberry — APPROVED** [review]
Cool! -- would you just add one comment and ping me to merge this, @pmeier?

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31111698).

**Pull request closed without merging**
